python/ray/data/_internal/execution/streaming_executor_state.py

```diff
@@ -330,6 +339,47 @@ def select_operator_to_run(
     )
 
 
+def _try_to_scale_up_cluster(topology: Topology):
+    """Try to scale up the cluster to accomodate the provided in-progress workload.
+
+    This makes a resource request to Ray's autoscaler consisting of the current,
+    aggregate usage of all operators in the DAG + the incremental usage of all operators
+    that are ready for dispatch (i.e. that have inputs queued). If the autoscaler were
+    to grant this resource request, it would allow us to dispatch one task for every
+    ready operator.
+
+    Note that this resource request does not take the global resource limits or the
```

Comment:
Hmm so it won't work if there are multiple concurrent streaming executions on the cluster right

Reply:
This global limit is already taken into account by the `_execution_allowed` before this function (it'll only be invoked when there is resource constraint).
For multiple executions, the situation seems the same as multiple datasets running with same limits, and they will each use up to that limits. If we settle on leaving limits at Execution Options, this will be the case.

Reply:
What I mean is the request_resources() will overwrite each other, which will break autoscaling entirely. You want the requests to be summed before sent to the autoscaler I think.

Reply:
I see, I didn't realize the calls with overwrite prior calls. How should this interact with autoscaler then?
- From API, it says the request_resources() is "only a hint to the autoscaler": so we may invoke autoscaling even if we don't have to (when usage is close to but still under the limits)?
- For the 'bundles` arg, it tries to "ensure this set of resource shapes can fit": does it try to find one node to fit each element in `bundles' (like placement group bundle)? If so, the current implementation is not correct.
- For summing resources before request resources, is summing up across operators not enough? Do you mean we should sum across different executor instances that are concurrent (how this can be done)?

Reply:
> For summing resources before request resources, is summing up across operators not enough? Do you mean we should sum across different executor instances that are concurrent (how this can be done)?

Yes, this is the main thing. I think a reasonably safe way to do this is to have a named actor that submits a running `sum` of requests to the autoscaler (e.g., sum of resource requests it has seen in the last 30 seconds or so). We may need to tweak the local algorithm to always submit scaling requests periodically to ensure the autoscaling.

Alternatively we could submit the running `max`, which I think would also ensure autoscaling, but may not be as good as the sum.